<launch>
  <arg name="mav_name" default="red"/>

  <!-- Size of Challenge, in meters (ICUAS-Simulation) -->
  <arg name="map_size_x" default="25.0"/>
  <arg name="map_size_y" default="16.0"/>
  <arg name="map_size_z" default="5.0"/>

  <arg name="odom_topic" default="$(arg mav_name)/odometry" />
  <arg name="camera_pose_topic" default="$(arg mav_name)/camera/pose"/>
  <arg name="depth_topic" default="$(arg mav_name)/camera/depth/image_raw"/>

  <!-- main algorithm params -->
  <include file="$(find ego_planner)/launch/advanced_param.xml">

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>
    <arg name="odometry_topic" value="$(arg odom_topic)"/>

    <!-- camera pose: transform of camera frame in the world frame -->
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- don't set cloud_topic if you already set these ones! -->
    <arg name="camera_pose_topic" value="$(arg camera_pose_topic)"/>
    <arg name="depth_topic" value="$(arg depth_topic)"/>

    <!-- topic of point cloud measurement, such as from LIDAR  -->
    <!-- don't set camera pose and depth, if you already set this one! -->
    <arg name="cloud_topic" value="/cloud"/>

    <!-- intrinsic params of the depth camera (ICUAS-Simulation)-->
    <arg name="cx" value="320.5"/>
    <arg name="cy" value="240.5"/>
    <arg name="fx" value="381.36246688113556"/>
    <arg name="fy" value="381.36246688113556"/>

    <!-- Max Vel and ACC (topp_config_custom.yaml) -->
    <arg name="max_vel" default="5" />
    <arg name="max_acc" default="2.75" />

    <!--always set to 1.5 times grater than sensing horizen-->
    <arg name="planning_horizon" value="7.5" /> 

    <!-- 1: use 2D Nav Goal to select goal  -->
    <!-- 2: use global waypoints below  -->
    <arg name="flight_type" value="1" />
    
    <!-- global waypoints -->
    <!-- It generates a piecewise min-snap traj passing all waypoints -->
    <arg name="point_num" value="5" />

    <arg name="point0_x" value="-15.0" />
    <arg name="point0_y" value="0.0" />
    <arg name="point0_z" value="1.0" />

    <arg name="point1_x" value="0.0" />
    <arg name="point1_y" value="15.0" />
    <arg name="point1_z" value="1.0" />

    <arg name="point2_x" value="15.0" />
    <arg name="point2_y" value="0.0" />
    <arg name="point2_z" value="1.0" />

    <arg name="point3_x" value="0.0" />
    <arg name="point3_y" value="-15.0" />
    <arg name="point3_z" value="1.0" />

    <arg name="point4_x" value="-15.0" />
    <arg name="point4_y" value="0.0" />
    <arg name="point4_z" value="1.0" />
    
  </include>

  <include file="$(find ego_planner)/launch/rviz.launch" />
  
  <!-- trajectory server -->
  <node pkg="ego_planner" name="traj_server" type="traj_server" output="screen">
    <remap from="/position_cmd" to="planning/pos_cmd"/>
    <remap from="/odom_world" to="$(arg odom_topic)"/>
    <param name="traj_server/time_forward" value="1.0" type="double"/>
  </node>

  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
    <remap from="~odom" to="$(arg odom_topic)"/>        
    <remap from="~goal" to="/move_base_simple/goal"/>
    <remap from="~traj_start_trigger" to="/traj_start_trigger" />
    <param name="waypoint_type" value="manual-lonely-waypoint"/>    
  </node>

  <node pkg="ego_planner" name="camera_pose_publisher" type="camera_pose_publisher.py" output="screen">
      <param name="parent_frame" value="world" />
      <param name="child_frame" value="$(arg mav_name)/camera" />
      <param name="pose_topic" value="$(arg mav_name)/camera/pose" />
  </node>

  <node pkg="ego_planner" name="trajectory_msg_converter" type="trajectory_msg_converter.py" output="screen">
      <param name="fast_planner_traj_topic" value="planning/pos_cmd"/>
      <param name="traj_pub_topic" value="$(arg mav_name)/position_hold/trajectory"/>
  </node>
</launch>
